Remove wrapDispatch and add mapMsgWithModel

bender2k14 opened the pull request:
This PR is a merge into the new branch `v4`.  It replaces the draft PR #244.

@cmeeren, there are conflicts between this branch and your logging branch.  Do you mind if we review and complete this PR first and then you resolve the conflicts with your logging branch?

cmeeren commented:
> @cmeeren, there are conflicts between this branch and your logging branch. Do you mind if we review and complete this PR first and then you resolve the conflicts with your logging branch?

Let's do!

cmeeren commented:
Is there anything outstanding here? Anything I should review that I haven't already?

cmeeren commented:
Come to think of it, IMHO the PRs should include relevant documentation/README updates, such as changelog items, migration steps, and relevant changes to the tutorial.

bender2k14 commented:
I updated the release notes.  What other documentation changes would you like ot see?

cmeeren commented on RELEASE_NOTES.md:
```diff
@@ -1,3 +1,11 @@
+#### 4.0.0
+
+* **Breaking:** Removed all occurrences of the argument `wrapDispatch` from the methods used to create a binding.  There is currently no migration path.  Please create an issue if this is a negative impact for you.
```
For consistency, please have only one space between sentences.

bender2k14 replied:
Done

cmeeren commented:
Please remove the "Wrapping dispatch" section of the tutorial.

I'd also like the tutorial to explain and demonstrate the new functions. Where/when are they useful, how to use them correctly, any gotchas to watch out for if applicable, etc. Simply put, make users who have no idea they need these functions understand how great they are. :)

bender2k14 commented:
I updated the tutorial.  How does that look?

bender2k14 commented:
> Is there anything outstanding here?

Another thing to review is all the types I exposed in [this commit](https://github.com/elmish/Elmish.WPF/pull/256/commits/63d889d88d79403c385af46306867ef7e5eabfdf).  Should I add `///`-documentation to all these types?

cmeeren commented:
> I updated the tutorial. How does that look?

Looks good! Could you also add something about how it enables you to factor out XAML controls? That's a (perhaps *the*?) major benefit, as I understand it.


> Another thing to review is all the types I exposed in [this commit](https://github.com/elmish/Elmish.WPF/pull/256/commits/63d889d88d79403c385af46306867ef7e5eabfdf). Should I add `///`-documentation to all these types?

Actually, could you make the implementations internal? `type X = internal { ... }`. I see no reason to make the internals of those types public.

Actually, why are they public in the first place?

bender2k14 commented:
> Could you also add something about how it enables you to factor out XAML controls? That's a (perhaps _the_?) major benefit, as I understand it.

The XAML can already be extracted.  For the `SubModelSeq` sample, I extracted it about six weeks ago in [this commit](https://github.com/elmish/Elmish.WPF/commit/d81c1ed140178f6e43805f06d8c150ec22833515) that is already in `master`.  The point related to the XAML is that it is now simpler and easier to create a design-time view model for the extracted XAML.  For the `SubModelSeq` sample, I did that in [this commit](https://github.com/elmish/Elmish.WPF/pull/256/commits/06a950927d7204be82d76d864cea4afb2f911017) in the of this PR.

I will add something about this to the tutorial.

> Actually, why are they public in the first place?

Ah, yes!  You are right.  Thanks for pushing back on this.  I wasn't being very precise when I did that commit.  Conceptually what I wanted was to publicly expose the six mapping functions (three in `Bindings` module and three in the `Binding` module).  I carelessly did that by only removing `internal` from the module, which was necessary, but the `Binding` module also contained `mapData` and `subModelSelectedItemLast`.  to publicly expose those required making the more things public.

Instead, I just rebased into my branch [this commit](https://github.com/elmish/Elmish.WPF/pull/256/commits/f8a35ed389d3e1b93160ddab78abc3ef98c2bab1) that publicly exposes the six mapping functions but keeps everything else internal.

So I will add `///`-style documentation comments to those six functions.

cmeeren commented on TUTORIAL.md:
```diff
@@ -775,7 +775,11 @@ let parentBindings () : Binding<ParentModel, ParentMsg> list =
+
+With such duplicate mapping code extracted, it is easier to create a design-time view model for the XAML code containing the bindings to `GrandChild1` and `GrandChild2`.  Specifically, instead of creating the design-time view model from the `parentBindings` bindings, it can now be created from the `childBindings` bindings.
+
+The `SubModelSeq` sample uses this benefit to create a design-time view model for `Counter.XAML`.  It also contains an example use of `mapModelWithMsg`.
```
`Counter.xaml` (lowercase extension)

bender2k14 replied:
done

cmeeren commented on src/Elmish.WPF/Binding.fs:
```diff
@@ -253,19 +253,53 @@ module Binding =
+  /// </summary>
+  /// <param name="f">The mapping function.</param>
+  /// <param name="binding">The binding to be mapped.</param>
+  let mapModel (f: 'a -> 'b) (binding: Binding<'b, 'msg>) = binding |> mapData (BindingData.mapModel f)
```
Regarding the docs you've added to these functions: I don't think Intellisense will ever show the param docs for F# functions. You can type `(` after the function name to see the first one, but I have never been able to make the rest show up. (For methods it works fine.)

I usually don't document function parameters unless there's something important to be said, and if so, integrate it into the main text.

For that reason, I also never use `<summary>`; I just do

```f#
/// Paragraph 1
///
/// Paragraph 2
let myFun = ...
```

bender2k14 replied:
Ok.  Actually thought that about that, but wrote it this way anyway.  I think descriptions for the arguments are not needed anyway.

I pushed [this commit](https://github.com/elmish/Elmish.WPF/pull/256/commits/ad8e85bdae367fd12c84a387325ea73cac9ed7c0) to make this change.  I will leave this for you to resolve or comment again.

bender2k14 commented:
As you have noticed (but repeating here for the record), I have added to the tutorial about the design-time benefit that comes from the mapping functions and added `///`-style documentation (and good type annotations and explicitly accepted all arguments with good names).

cmeeren commented:
Looks good to me. Feel free to merge!

bender2k14 commented:
> Looks good to me. Feel free to merge!

Great! :D

I squashed some commits.  I will complete after the automated build finishes successfully.
